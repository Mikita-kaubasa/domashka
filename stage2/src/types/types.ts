/**
 * @fileoverview TypeScript type definitions for the user management application.
 * This module contains all the core interfaces and types used throughout the application,
 * including user data structures, address information, and company details.
 * 
 * @module types
 */

/**
 * Geographic coordinates for a location.
 * 
 * @interface Geo
 * @property {string} lat - Latitude coordinate as a string
 * @property {string} lng - Longitude coordinate as a string
 * 
 * @example
 * ```typescript
 * const coordinates: Geo = {
 *   lat: "40.7128",
 *   lng: "-74.0060"
 * };
 * ```
 */
export interface Geo {
  lat: string;
  lng: string;
}

/**
 * Complete address information for a user.
 * 
 * @interface Address
 * @property {string} street - Street name and number
 * @property {string} suite - Suite, apartment, or unit number
 * @property {string} city - City name
 * @property {string} zipcode - Postal/ZIP code
 * @property {Geo} geo - Geographic coordinates of the address
 * 
 * @example
 * ```typescript
 * const address: Address = {
 *   street: "123 Main St",
 *   suite: "Apt 4B",
 *   city: "New York",
 *   zipcode: "10001",
 *   geo: { lat: "40.7128", lng: "-74.0060" }
 * };
 * ```
 */
export interface Address {
  street: string;
  suite: string;
  city: string;
  zipcode: string;
  geo: Geo;
}

/**
 * Company information associated with a user.
 * 
 * @interface Company
 * @property {string} name - Company name
 * @property {string} catchPhrase - Company slogan or catchphrase
 * @property {string} bs - Business strategy or description
 * 
 * @example
 * ```typescript
 * const company: Company = {
 *   name: "Acme Corp",
 *   catchPhrase: "Making the world a better place",
 *   bs: "harness real-time e-markets"
 * };
 * ```
 */
export interface Company {
  name: string;
  catchPhrase: string;
  bs: string;
}

/**
 * Complete user information structure.
 * 
 * This is the main data structure representing a user in the application.
 * It contains all personal, contact, and professional information.
 * 
 * @interface User
 * @property {number} id - Unique identifier for the user
 * @property {string} name - Full name of the user
 * @property {string} username - Unique username for the user
 * @property {string} email - Email address of the user
 * @property {Address} address - Complete address information
 * @property {string} phone - Phone number (can be in various formats)
 * @property {string} website - Personal or company website URL
 * @property {Company} company - Associated company information
 * 
 * @example
 * ```typescript
 * const user: User = {
 *   id: 1,
 *   name: "John Doe",
 *   username: "johndoe",
 *   email: "john@example.com",
 *   address: {
 *     street: "123 Main St",
 *     suite: "Apt 4B",
 *     city: "New York",
 *     zipcode: "10001",
 *     geo: { lat: "40.7128", lng: "-74.0060" }
 *   },
 *   phone: "1-770-736-8031 x56442",
 *   website: "johndoe.org",
 *   company: {
 *     name: "Acme Corp",
 *     catchPhrase: "Making the world a better place",
 *     bs: "harness real-time e-markets"
 *   }
 * };
 * ```
 */
export interface User {
  id: number;
  name: string;
  username: string;
  email: string;
  address: Address;
  phone: string;
  website: string;
  company: Company;
} 