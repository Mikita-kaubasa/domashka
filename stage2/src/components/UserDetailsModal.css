.modal-backdrop {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background: rgba(0, 0, 0, 0.25);
    display: flex;
    align-items: center;
    justify-content: center;
    z-index: 1000;
    padding: 1rem;
    animation: fadeIn 0.3s ease-out;
}

@keyframes fadeIn {
    from {
        opacity: 0;
    }

    to {
        opacity: 1;
    }
}

.modal-container.redesigned-modal {
    background: #fff;
    border-radius: 20px;
    max-width: 700px;
    width: 100%;
    box-shadow: 0 8px 32px 0 rgba(0, 0, 0, 0.18);
    animation: slideIn 0.3s ease-out;
    display: flex;
    flex-direction: column;
    padding: 0;
}

@keyframes slideIn {
    from {
        transform: translateY(-20px) scale(0.97);
        opacity: 0;
    }

    to {
        transform: translateY(0) scale(1);
        opacity: 1;
    }
}

.redesigned-header {
    display: flex;
    justify-content: space-between;
    align-items: flex-start;
    padding: 2.1rem 3.2rem 1.2rem 3.2rem;
    background: transparent;
    border-radius: 20px 20px 0 0;
    text-align: left;
}

.modal-user-name {
    font-size: 1.55rem;
    font-weight: 700;
    color: #222;
    margin-bottom: 0.3rem;
    line-height: 1.2;
    text-align: left;
}

.modal-user-email {
    color: #2563eb;
    font-size: 1.08rem;
    text-decoration: none;
    font-weight: 500;
    word-break: break-all;
    text-align: left;
    display: block;
}

.redesigned-close {
    background: none;
    border: none;
    color: #222;
    font-size: 1.5rem;
    font-weight: 400;
    cursor: pointer;
    padding: 0.2rem 0.5rem;
    margin-left: 1rem;
    border-radius: 6px;
    transition: background 0.15s;
    line-height: 1;
}

.redesigned-close:hover,
.redesigned-close:focus {
    background: #f3f4f6;
}

.modal-divider {
    width: 100%;
    height: 1px;
    background: #ececec;
    margin: 0;
}

.redesigned-content {
    padding: 1.5rem 3.2rem 2.1rem 3.2rem;
    display: flex;
    flex-direction: column;
    gap: 2.1rem;
    text-align: left;
}

.modal-section {
    margin-bottom: 0;
}

.modal-section-title {
    font-size: 1.08rem;
    font-weight: 600;
    color: #222;
    margin-bottom: 0.7rem;
    letter-spacing: 0.01em;
    text-align: left;
}

.modal-section-body {
    display: flex;
    flex-direction: column;
    gap: 0.3rem;
    color: #222;
    font-size: 1.01rem;
    text-align: left;
}

.modal-label {
    font-weight: 600;
    color: #222;
    margin-right: 0.2em;
    text-align: left;
}

.modal-map-link {
    color: #2563eb;
    font-weight: 500;
    text-decoration: none;
    margin-top: 0.5rem;
    display: inline-flex;
    align-items: center;
    font-size: 1.01rem;
    transition: color 0.15s;
    text-align: left;
}

.modal-map-link:hover {
    color: #1d4ed8;
    text-decoration: underline;
}

.modal-map-pin {
    margin-right: 0.3em;
    font-size: 1.1em;
}

.modal-link {
    color: #2563eb;
    font-weight: 500;
    text-decoration: none;
    transition: color 0.15s;
}

.modal-link:hover {
    color: #1d4ed8;
    text-decoration: underline;
}

@media (max-width: 900px) {
    .modal-container.redesigned-modal {
        max-width: 98vw;
        padding: 0;
    }

    .redesigned-header,
    .redesigned-content {
        padding-left: 1.2rem;
        padding-right: 1.2rem;
    }

    .redesigned-header {
        padding-top: 1.2rem;
        padding-bottom: 0.7rem;
    }

    .redesigned-content {
        padding-bottom: 1.2rem;
        gap: 1.2rem;
    }
}