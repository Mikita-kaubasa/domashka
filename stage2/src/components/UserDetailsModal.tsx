/**
 * @fileoverview Modal component for displaying detailed user information.
 * This component provides a comprehensive view of user details including contact information,
 * address, and company details in a modal overlay with accessibility features.
 * 
 * @module UserDetailsModal
 * @requires React
 * @requires ../types/types
 * @requires ../utils/formatters
 * @requires ./UserDetailsModal.css
 */

import React, { memo, useCallback, useEffect, useMemo, useRef } from 'react';
import { User } from '../types/types';
import { formatAddress, formatWebsite, getMapUrl } from '../utils/formatters';
import './UserDetailsModal.css';

/**
 * Props for the UserDetailsModal component.
 * 
 * @interface UserDetailsModalProps
 * @property {User | null} user - The user object to display, null if no user selected
 * @property {boolean} isOpen - Whether the modal is currently open
 * @property {() => void} onClose - Function to call when the modal should be closed
 */
interface UserDetailsModalProps {
    user: User | null;
    isOpen: boolean;
    onClose: () => void;
}

/**
 * Modal component for displaying detailed user information.
 * 
 * This component provides:
 * - Comprehensive user details display
 * - Modal overlay with backdrop
 * - Keyboard navigation (Escape to close)
 * - Click outside to close functionality
 * - Body scroll prevention when open
 * - Auto-focus management
 * - Accessibility features (ARIA labels, keyboard support)
 * - Responsive design with modern styling
 * 
 * @component
 * @param {UserDetailsModalProps} props - Component props
 * @returns {JSX.Element | null} Modal component or null if not open
 * 
 * @example
 * ```tsx
 * import UserDetailsModal from './components/UserDetailsModal';
 * 
 * function App() {
 *   const [selectedUser, setSelectedUser] = useState<User | null>(null);
 *   const [isModalOpen, setIsModalOpen] = useState(false);
 * 
 *   return (
 *     <UserDetailsModal
 *       user={selectedUser}
 *       isOpen={isModalOpen}
 *       onClose={() => setIsModalOpen(false)}
 *     />
 *   );
 * }
 * ```
 */
const UserDetailsModal: React.FC<UserDetailsModalProps> = memo(({ user, isOpen, onClose }) => {
    const modalRef = useRef<HTMLDivElement>(null);

    /**
     * Prevents body scrolling when modal is open and restores it when closed.
     * 
     * This effect ensures that the background content doesn't scroll while
     * the modal is open, providing a better user experience.
     * 
     * @example
     * ```typescript
     * // Body scroll is automatically managed when modal opens/closes
     * ```
     */
    useEffect(() => {
        if (isOpen) {
            document.body.style.overflow = 'hidden';
        } else {
            document.body.style.overflow = 'unset';
        }

        // Cleanup function to restore scrolling when component unmounts
        return () => {
            document.body.style.overflow = 'unset';
        };
    }, [isOpen]);

    /**
     * Auto-focuses the modal when it opens for better accessibility.
     * 
     * This ensures that keyboard users can immediately interact with the modal
     * when it opens, following accessibility best practices.
     * 
     * @example
     * ```typescript
     * // Modal automatically receives focus when opened
     * ```
     */
    useEffect(() => {
        if (isOpen && modalRef.current) {
            modalRef.current.focus();
        }
    }, [isOpen]);

    /**
     * Handles backdrop click to close the modal.
     * 
     * Only closes the modal if the click target is the backdrop itself,
     * not if the click is on the modal content.
     * 
     * @param {React.MouseEvent} e - The mouse event
     * 
     * @example
     * ```typescript
     * // Clicking outside the modal content closes it
     * ```
     */
    const handleBackdropClick = useCallback((e: React.MouseEvent) => {
        if (e.target === e.currentTarget) {
            onClose();
        }
    }, [onClose]);

    /**
     * Handles keyboard events for accessibility.
     * 
     * Supports Escape key to close the modal, following standard
     * modal interaction patterns.
     * 
     * @param {React.KeyboardEvent} e - The keyboard event
     * 
     * @example
     * ```typescript
     * // Pressing Escape closes the modal
     * ```
     */
    const handleKeyDown = useCallback((e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
            onClose();
        }
    }, [onClose]);

    /**
     * Memoized formatted address string.
     * 
     * Only recalculates when the user object changes to optimize performance.
     * 
     * @type {string}
     */
    const formattedAddress = useMemo(() =>
        user ? formatAddress(user.address) : '',
        [user]
    );

    /**
     * Memoized formatted website URL.
     * 
     * Ensures the website URL has proper protocol for external links.
     * 
     * @type {string}
     */
    const formattedWebsite = useMemo(() =>
        user ? formatWebsite(user.website) : '',
        [user]
    );

    /**
     * Memoized Google Maps URL for the user's location.
     * 
     * Creates a clickable link to view the user's address on Google Maps.
     * 
     * @type {string}
     */
    const mapUrl = useMemo(() =>
        user ? getMapUrl(user.address.geo.lat, user.address.geo.lng) : '',
        [user]
    );

    // Don't render if modal is not open or no user is selected
    if (!isOpen || !user) return null;

    return (
        <div
            ref={modalRef}
            className="modal-backdrop"
            onClick={handleBackdropClick}
            onKeyDown={handleKeyDown}
            tabIndex={-1}
        >
            <div className="modal-container redesigned-modal">
                <div className="modal-header redesigned-header">
                    <div>
                        <div className="modal-user-name">{user.name}</div>
                        <a href={`mailto:${user.email}`} className="modal-user-email">{user.email}</a>
                    </div>
                    <button className="close-button redesigned-close" onClick={onClose} aria-label="Close modal">
                        <span aria-hidden="true">√ó</span>
                    </button>
                </div>
                <div className="modal-divider" />
                <div className="modal-content redesigned-content">
                    <div className="modal-section">
                        <div className="modal-section-title">Address</div>
                        <div className="modal-section-body">
                            <div>{user.address.street}, {user.address.suite}</div>
                            <div>{user.address.city}, {user.address.zipcode}</div>
                            <a
                                href={mapUrl}
                                target="_blank"
                                rel="noopener noreferrer"
                                className="modal-map-link"
                            >
                                <span className="modal-map-pin" role="img" aria-label="pin">üìç</span> View on map
                            </a>
                        </div>
                    </div>
                    <div className="modal-section">
                        <div className="modal-section-title">Contact</div>
                        <div className="modal-section-body">
                            <div><span className="modal-label">Phone:</span> {user.phone}</div>
                            <div>
                                <span className="modal-label">Website:</span>{' '}
                                <a
                                    href={formattedWebsite}
                                    target="_blank"
                                    rel="noopener noreferrer"
                                    className="modal-link"
                                >
                                    {user.website}
                                </a>
                            </div>
                        </div>
                    </div>
                    <div className="modal-section">
                        <div className="modal-section-title">Company</div>
                        <div className="modal-section-body">
                            <div><span className="modal-label">Name:</span> {user.company.name}</div>
                            <div><span className="modal-label">Catchphrase:</span> {user.company.catchPhrase}</div>
                            <div><span className="modal-label">Business:</span> {user.company.bs}</div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
});

UserDetailsModal.displayName = 'UserDetailsModal';

export default UserDetailsModal; 