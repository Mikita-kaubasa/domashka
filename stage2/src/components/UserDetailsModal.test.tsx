import { fireEvent, render, screen } from '@testing-library/react';
import { User } from '../types/types';
import UserDetailsModal from './UserDetailsModal';

const mockUser: User = {
    id: 1,
    name: 'Alice',
    username: 'alice123',
    email: 'alice@example.com',
    address: {
        street: 'Main St',
        suite: 'Apt 1',
        city: 'Wonderland',
        zipcode: '12345',
        geo: { lat: '0', lng: '0' },
    },
    phone: '123-456-7890',
    website: 'alice.com',
    company: {
        name: 'Alice Co',
        catchPhrase: 'Curious and Curiouser',
        bs: 'adventures',
    },
};

describe('UserDetailsModal', () => {
    it('renders without crashing when open and user is provided', () => {
        render(<UserDetailsModal user={mockUser} isOpen={true} onClose={jest.fn()} />);
        expect(screen.getByText('Alice')).toBeInTheDocument();
        expect(screen.getByText('alice@example.com')).toBeInTheDocument();
        expect(screen.getByText(/Address/i)).toBeInTheDocument();
        expect(screen.getByText(/Contact/i)).toBeInTheDocument();
        expect(screen.getByText(/Company/i)).toBeInTheDocument();
    });

    it('does not render when isOpen is false', () => {
        render(<UserDetailsModal user={mockUser} isOpen={false} onClose={jest.fn()} />);
        expect(screen.queryByText('Alice')).not.toBeInTheDocument();
    });

    it('does not render when user is null', () => {
        render(<UserDetailsModal user={null} isOpen={true} onClose={jest.fn()} />);
        expect(screen.queryByText(/Address/i)).not.toBeInTheDocument();
    });

    it('calls onClose when close button is clicked', () => {
        const handleClose = jest.fn();
        render(<UserDetailsModal user={mockUser} isOpen={true} onClose={handleClose} />);
        const closeButton = screen.getByRole('button', { name: /close modal/i });
        fireEvent.click(closeButton);
        expect(handleClose).toHaveBeenCalled();
    });

    it('calls onClose when Escape key is pressed', () => {
        const handleClose = jest.fn();
        render(<UserDetailsModal user={mockUser} isOpen={true} onClose={handleClose} />);
        const modal = screen.getByText('Alice').closest('.modal-backdrop');
        if (modal) {
            fireEvent.keyDown(modal, { key: 'Escape', code: 'Escape' });
            expect(handleClose).toHaveBeenCalled();
        }
    });

    it('has correct accessibility attributes', () => {
        render(<UserDetailsModal user={mockUser} isOpen={true} onClose={jest.fn()} />);
        const modal = screen.getByText('Alice').closest('.modal-backdrop');
        expect(modal).toHaveAttribute('tabindex', '-1');
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<UserDetailsModal user={mockUser} isOpen={true} onClose={jest.fn()} />);
        expect(asFragment()).toMatchSnapshot();
    });
}); 