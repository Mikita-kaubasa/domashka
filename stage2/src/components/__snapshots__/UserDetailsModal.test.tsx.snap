// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserDetailsModal matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal-backdrop"
    tabindex="-1"
  >
    <div
      class="modal-container redesigned-modal"
    >
      <div
        class="modal-header redesigned-header"
      >
        <div>
          <div
            class="modal-user-name"
          >
            Alice
          </div>
          <a
            class="modal-user-email"
            href="mailto:alice@example.com"
          >
            alice@example.com
          </a>
        </div>
        <button
          aria-label="Close modal"
          class="close-button redesigned-close"
        >
          <span
            aria-hidden="true"
          >
            √ó
          </span>
        </button>
      </div>
      <div
        class="modal-divider"
      />
      <div
        class="modal-content redesigned-content"
      >
        <div
          class="modal-section"
        >
          <div
            class="modal-section-title"
          >
            Address
          </div>
          <div
            class="modal-section-body"
          >
            <div>
              Main St, Apt 1
            </div>
            <div>
              Wonderland, 12345
            </div>
            <a
              class="modal-map-link"
              href="https://www.google.com/maps?q=0,0"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                aria-label="pin"
                class="modal-map-pin"
                role="img"
              >
                üìç
              </span>
               View on map
            </a>
          </div>
        </div>
        <div
          class="modal-section"
        >
          <div
            class="modal-section-title"
          >
            Contact
          </div>
          <div
            class="modal-section-body"
          >
            <div>
              <span
                class="modal-label"
              >
                Phone:
              </span>
               123-456-7890
            </div>
            <div>
              <span
                class="modal-label"
              >
                Website:
              </span>
               
              <a
                class="modal-link"
                href="https://alice.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                alice.com
              </a>
            </div>
          </div>
        </div>
        <div
          class="modal-section"
        >
          <div
            class="modal-section-title"
          >
            Company
          </div>
          <div
            class="modal-section-body"
          >
            <div>
              <span
                class="modal-label"
              >
                Name:
              </span>
               Alice Co
            </div>
            <div>
              <span
                class="modal-label"
              >
                Catchphrase:
              </span>
               Curious and Curiouser
            </div>
            <div>
              <span
                class="modal-label"
              >
                Business:
              </span>
               adventures
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
